libyul/optimiser/LabelIDDispenser.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/LabelIDDispenser.h>
+
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+
+#include <fmt/compile.h>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/iota.hpp>
+
+using namespace solidity::yul;
+
+namespace
+{
+bool isInvalidLabel(
+	std::string_view const _label,
+	std::set<std::string, std::less<>> const& _reservedLabels,
+	Dialect const& _dialect
+)
+{
+	return isRestrictedIdentifier(_dialect, _label) || _reservedLabels.contains(_label);
+}
+}
+
+LabelIDDispenser::LabelIDDispenser(ASTLabelRegistry const& _labels, std::set<std::string> const& _reservedLabels):
+	m_labels(_labels),
+	m_reservedLabels(_reservedLabels.begin(), _reservedLabels.end()),
+	m_offset(_labels.maxID() + 1)
+{
+	m_reservedLabels += std::set{""};
```

Comment:
Is it even necessary to reserve `""`? As far as I can tell, `m_reservedLabels` just prevents a label from being reused in `generateNewLabels()`, but you set `reusedLabels[0] = true` there, so it won't hit the reserved name check anyway.

Reply:
Good catch. I don't think it is necessary and having fewer reserved labels is good for performance. :)